lib/msal-node/src/cache/distributed/DistributedCachePlugin.ts

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { AccountEntity, ICachePlugin, TokenCacheContext } from "@azure/msal-common";
+import { TokenCache } from "../TokenCache";
+import { IPartitionManager } from "./IPartitionManager";
+import { ICacheClient } from "./ICacheClient";
+
+export class DistributedCachePlugin implements ICachePlugin {
+    private client: ICacheClient;
+    private partitionManager?: IPartitionManager;
+
+    constructor(client: ICacheClient, partitionManager?: IPartitionManager) {
+        this.client = client;
+        this.partitionManager = partitionManager;
+    }
+  
+    public async beforeCacheAccess(cacheContext: TokenCacheContext): Promise<void> {
+        if (this.partitionManager) {
+            const partitionKey = await this.partitionManager.getKey();
+            const cacheData = await this.client.get(partitionKey);
+            cacheContext.tokenCache.deserialize(cacheData);
+        }
+    }
+  
+    public async afterCacheAccess(cacheContext: TokenCacheContext): Promise<void> {
+        if (cacheContext.cacheHasChanged && this.partitionManager) {
+            const kvStore = (cacheContext.tokenCache as TokenCache).getKVStore();
+  
+            if (Object.keys(kvStore).length >= 2) {
+                const accountEntity = Object.values(kvStore)[1] as AccountEntity; // the second entity is the account
+                const partitionKey = await this.partitionManager.extractKey(accountEntity);
+                          
+                await this.client.set(partitionKey, cacheContext.tokenCache.serialize());           
+            }
+        }
+    }
```

Comment:
This strikes me as not a sufficient api for a distributed scenarios. I believe you need to alter this api to include hooks for intercepting when individual keys are accessed rather than the entire cache.

I don't believe that loading and persisting the entire cache before and after every access is scalable. If I have millions of users I cannot have the entire object stored in memory and I can't load it all on every access and the process of just reading and writing that document in a horizontally scaled environment will cause various servers to get out of sync.

Instead I would _add_ additional functions to implement that would let you hook in before/after each individual key is accessed. That would allow me to manage fetching and persisting only individual records used by each server on demand.

For example it might look like this:

```ts
export class DistributedCachePlugin implements ICachePlugin {
  public async beforeCacheKeyAccess(cacheContext: TokenCacheContext, key: string): Promise<void> {
    // if not in memory, or if whats in memory is expried...
    // fetch individual record from database using key
    // if in db and not expired then load it into in memory cache
  }

  public async afterCacheKeyAccess(cacheContext: TokenCacheContext, string key): Promise<void> {
    // grab record from memory by key
    // persist into DB
  }
}
```

And these functions would be called before / after accessing the records stored on the kv store itself. Or optionally, make an IStore interface and allow that kvstore to be a plugin.

I just don't think its reasonable to attempt to serialize and deserialize and entire token cache at all, and you probably need to abstract the store itself to be a database just load it all into memory and persist it into a database.

Reply:
To elaborate on that last point, probably what you really need is to make the storage module plugins rather than a hard coded instance in the ClientApplication constructor:

https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/src/client/ClientApplication.ts#L77-L87

```ts
export abstract class ClientApplication {
  // private fields...
  protected storage: IStoragePlugin;

  protected constructor(configuration: Configuration) {
    this.config = buildAppConfiguration(configuration);
    this.cryptoProvider = new CryptoProvider();
    this.logger = new Logger(this.config.system.loggerOptions, name, version);
    this.storage = configuration.storagePlugin(this.logger, this.cryptoProvider) ?? new NodeStorage(this.logger, this.config.auth.clientId, this.cryptoProvider);
    this.tokenCache = new TokenCache(
        this.storage,
        this.logger,
        this.config.cache.cachePlugin
    );
  }
}
```

Reply:
Hi @justinmchase,

Thank you for taking the time to review the pull request, I really appreciate that.
The partition manager in this case ensures that the specific instance is only pulling and storing into cache the information specific to the user in question. We are not loading the entire cache each time, rather we are loading the user specific information only for this instance. See line 22 and 34 for more on this.

We have also added a sample in `samples/msal-node-samples/ExpressTestApp/RedisTestApp` that shows this in action and from the sample you will not that we are using session data to store the specific key for the user and loading only their information into cache.

Reply:
Hi @justinmchase, I agree that we can go the StoragePlugin route, there are definitely two ways to skin this cat.

However, I want to better understand, why you think the proposed solution will have race conditions. And yes you are right, the access to the cache is being defined by the user of the plugin as they implement the `IPartitionManager` and thus handle how the cache, which ever it maybe is accessed in a safe way.

Reply:
When the request is first received, this line is invoked: 

`partitionManager.setSessionId(req.session.id);`

And then downstream from that, the plugin reads/write from the cache, assuming `partitionManager.setSessionId(req.session.id);` hasn't been called again with a different `req.session.id`, however, it you make multiple requests to the same server instance in parallel, it looks like there is a possibility that a request for one user would set the session id on the partition manager (which is a singleton in this sample), but then between the time that happens and the time the plugin is invoked and the data is retrieved from the cache, another request for another user could have update that session id to the second user, which would result in the first user getting data for the second user. 

I agree with the feedback that the plugin should get the session id explicitly as a parameter, so that way it doesn't rely upon global state.

Reply:
You are both right, I have refactored the sample to not share instances of the `PartitionManager` across calls by creating a new instance for each request as opposed to using the singleton we were using. I would love your review on the refactor.